edweiss-app/hooks/repository.tsx

```diff
@@ -0,0 +1,348 @@
+import { Document, DocumentData, Query, getDocuments } from '@/config/firebase';
+import { CallResult } from '@/model/functions';
+import generateUID from '@/utils/uid';
+import { Stack } from 'expo-router';
+import React, { ReactNode, useCallback, useContext, useEffect, useMemo, useRef } from 'react';
+import { useStoredState } from './storage';
+
+/**
+ * Encapsulates everything that is necessary to use a repository.
+ */
+export interface Repository<T extends DocumentData> {
+	/**
+	 * Unique key that is used when writing locally.
+	 */
+	readonly key: string,
+
+	/**
+	 * React context that is used when propagating the repository.
+	 */
+	readonly context: React.Context<RepositoryInstance<T>>
+}
+
+export interface RepositoryDocument<Type> extends Document<Type> {
+	/**
+	 * `true` if the `id` field of this object is the Firebase one.
+	 */
+	syncedId: boolean,
+
+	/**
+	 * Temporary id used when adding a document.
+	 */
+	readonly fakeId?: string
+}
+
+type MonoSyncCallback = (id: string) => void;
+type MultiSyncCallback = (id: string[]) => void;
+
+/**
+ * Used to manipulate and modify the local {@link Repository} and enforce synchronicity with the cloud.
+ */
+export interface RepositoryHandler<T extends DocumentData> {
+	/**
+	 * Adds a document to the local {@link Repository} then syncs it with the cloud using `idSupplier`.
+	 * @param data The data of the document to be added optimistically.
+	 * @param idSupplier Typically a `callFunction` with an `{ id: string }` as a success return value.
+	 */
+	readonly addDocument: (data: T, idSupplier: Promise<CallResult<{ id: string }, unknown>>) => void,
```

Comment:
I see that there are two different functions to add or update a document. I can see an issue with this: the DocumentReference<T>.set() function that we used in createQuizAttempt acted as both a creation tool (when there are no attempts at the designated path) or a modifying tool (when there was indeed an attempt). 
```typescript
export const createQuizAttempt = onSanitizedCall(QuizzesAttempts.Functions.createQuizAttempt, {
	courseId: Predicate.isNonEmptyString,
	path: Predicate.isNonEmptyString,
	quizId: Predicate.isNonEmptyString,
	quizAttempt: CustomPredicateQuiz.isValidQuizAttempt
}, async (userId, args) => {
	const ref = getDocumentRef(CollectionOf<QuizzesAttempts.QuizAttempt>(args.path), userId);

	await ref.set(args.quizAttempt);
	console.log("Document successfully updated!");
	return ok({ id: ref.id });
});
```
Do we now have to decouple those two cases?

Reply:
No, as the Quiz is not applicable in the case of the Repository.
Repository is only used for modules that are actively under the CRUD model, for example it can be used for :
 - Flashcards
 - Todos
 - Notifications
Quizzes and attempts do not fit the Repository purpose.

Reply:
Ahh, I see, thank you. Should the other members be immediately aware of the changes you made and use the handler this sprint, or do you believe the code is easily refactorable later?

Reply:
I encourage them to adopt it as early as possible, because 1. it will inevitably provide the best possible user experience, and 2. the tests do change a bit so it's better to make the transition quickly.

Reply:
Since your PR is not merged yet, I propose we refactor right after tomorrows meeting. Do you agree?

Reply:
It's as you guys want, I'm of course not forcing anyone to use it. Only people who want their module to be blazingly fast, responsive and pleasant should implement it.